ExpressionEvalTest.test547462_BooleanExpression failed with "Address already in use" in I20260414-1800

See: https://download.eclipse.org/eclipse/downloads/drops4/I20260414-1800/testresults/html/org.eclipse.jdt.debug.tests_ep440I-unit-macosx-x86_64-java21_macosx.cocoa.x86_64_21.html

```
	Cannot connect to VM

org.eclipse.core.runtime.CoreException: Cannot connect to VM
at org.eclipse.jdt.launching.AbstractVMRunner.abort(AbstractVMRunner.java:56)
at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:457)
at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1495)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1473)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1669)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1624)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1596)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1583)
at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
at java.base/java.lang.reflect.Method.invoke(Method.java:580)
at junit.framework.TestCase.runTest(TestCase.java:177)
at junit.framework.TestCase.runBare(TestCase.java:142)
at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3007)
at junit.framework.TestResult$1.protect(TestResult.java:122)
at junit.framework.TestResult.runProtected(TestResult.java:142)
at junit.framework.TestResult.run(TestResult.java:125)
at junit.framework.TestCase.run(TestCase.java:130)
at junit.framework.TestSuite.runTest(TestSuite.java:241)
at junit.framework.TestSuite.run(TestSuite.java:236)
at junit.framework.TestSuite.runTest(TestSuite.java:241)
at org.eclipse.jdt.debug.tests.DebugSuite$1.run(DebugSuite.java:61)
at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.BindException: Address already in use
at java.base/sun.nio.ch.Net.bind0(Native Method)
at java.base/sun.nio.ch.Net.bind(Net.java:577)
at java.base/sun.nio.ch.Net.bind(Net.java:566)
at java.base/sun.nio.ch.NioSocketImpl.bind(NioSocketImpl.java:636)
at java.base/java.net.ServerSocket.bind(ServerSocket.java:391)
at java.base/java.net.ServerSocket.<init>(ServerSocket.java:278)
at java.base/java.net.ServerSocket.<init>(ServerSocket.java:171)
at org.eclipse.jdi.internal.connect.SocketTransportService.startListening(SocketTransportService.java:290)
at org.eclipse.jdi.internal.connect.SocketTransportImpl.startListening(SocketTransportImpl.java:53)
at org.eclipse.jdi.internal.connect.SocketListeningConnectorImpl.startListening(SocketListeningConnectorImpl.java:141)
at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:335)
... 23 more
```

Workspace log has nothing useful: https://download.eclipse.org/eclipse/downloads/drops4/I20260414-1800/testresults/ep440I-unit-macosx-x86_64-java21_macosx.cocoa.x86_64_21/org.eclipse.jdt.debug.tests.AutomatedSuite.txt

```
01:07:12.155 SETUP ExpressionEvalTest.test547462_BooleanExpression

!ENTRY org.eclipse.jdt.debug.ui.tests 1 0 2026-04-15 01:07:12.176
!MESSAGE 

01:07:12.176 TDOWN ExpressionEvalTest.test547462_BooleanExpression
```

The port is set here:

```

"main" #1 [16792] prio=6 os_prio=0 cpu=4448.61ms elapsed=59.48s tid=0x00007f082401bd40 nid=16792 runnable  [0x00007f08287fc000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.jdt.launching.SocketUtil.findFreePort(SocketUtil.java:69)
        at org.eclipse.jdt.internal.launching.StandardVMDebugger.getCommandLine(StandardVMDebugger.java:169)
        at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:310)
        at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
        at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1495)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1473)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1669)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1624)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1596)
        at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1583)
        at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
        at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
```
```
	public static int findFreePort() {
		try (ServerSocket socket = new ServerSocket(0)) {
			return socket.getLocalPort();
		} catch (IOException e) {
		}
		return -1;
	}
```

Maybe its no longer free... I'm not sure what we can do about this.

Trying this:

```
diff --git a/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java b/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java
index 2ee24a21b..1705f7674 100644
--- a/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java
+++ b/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java
@@ -19,6 +19,7 @@ package org.eclipse.jdt.debug.tests;
 import java.io.File;
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.net.BindException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -135,6 +136,7 @@ import org.eclipse.jdt.internal.debug.eval.ast.engine.ASTEvaluationEngine;
 import org.eclipse.jdt.internal.debug.ui.BreakpointUtils;
 import org.eclipse.jdt.internal.debug.ui.IJDIPreferencesConstants;
 import org.eclipse.jdt.internal.debug.ui.JDIDebugUIPlugin;
+import org.eclipse.jdt.internal.launching.StandardVMDebugger;
 import org.eclipse.jdt.launching.IJavaLaunchConfigurationConstants;
 import org.eclipse.jdt.launching.IVMInstall;
 import org.eclipse.jdt.launching.IVMInstallChangedListener;
@@ -177,6 +179,7 @@ import junit.framework.TestCase;
 @SuppressWarnings("deprecation")
 public abstract class AbstractDebugTest extends TestCase implements  IEvaluationListener {
 
+       private static final int SOCKET_BIND_ERROR_MAX_RETRIES = 3;
        private static boolean setupFirstTest = false;
 
        public static final String MULTI_OUTPUT_PROJECT_NAME = "MultiOutput";
@@ -1490,12 +1493,24 @@ public abstract class AbstractDebugTest extends TestCase implements  IEvaluation
         *             if the event is never received.
         */
        protected Object launchAndWait(ILaunchConfiguration configuration, String mode, DebugEventWaiter waiter, boolean register) throws CoreException {
-               ILaunch launch;
-               try {
-                       launch = configuration.launch(mode, new TimeoutMonitor(DEFAULT_TIMEOUT), false, register);
-               } catch (Throwable t) {
-                       logProcessConsoleContents();
-                       throw t;
+               ILaunch launch = null;
+               for (int retry = 1; retry <= SOCKET_BIND_ERROR_MAX_RETRIES; ++retry) {
+                       try {
+                               StandardVMDebugger.THROW = retry < SOCKET_BIND_ERROR_MAX_RETRIES;
+                               launch = configuration.launch(mode, new TimeoutMonitor(DEFAULT_TIMEOUT), false, register);
+                               break;
+                       } catch (CoreException e) {
+                               Throwable cause = e.getStatus().getException();
+                               if (retry < SOCKET_BIND_ERROR_MAX_RETRIES && cause instanceof BindException) {
+                                       // port might be in use between checking for a free port and launching, try a few times
+                                       DebugPlugin.log(Status.error("Could not bind socket for debugging, retry: " + retry, e));
+                               } else {
+                                       throw e;
+                               }
+                       } catch (Throwable t) {
+                               logProcessConsoleContents();
+                               throw t;
+                       }
                }
                Object suspendee= waiter.waitForEvent();
                if (suspendee == null) {
diff --git a/org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/StandardVMDebugger.java b/org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/StandardVMDebugger.java
index 7df5298a8..764cb3f87 100644
--- a/org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/StandardVMDebugger.java
+++ b/org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/StandardVMDebugger.java
@@ -18,6 +18,7 @@ package org.eclipse.jdt.internal.launching;
 import java.io.File;
 import java.io.IOException;
 import java.io.InterruptedIOException;
+import java.net.BindException;
 import java.text.DateFormat;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -298,6 +299,7 @@ public class StandardVMDebugger extends StandardVMRunner {
 
        }
 
+       public static boolean THROW = false;
        /*
         * (non-Javadoc)
         *
@@ -332,6 +334,9 @@ public class StandardVMDebugger extends StandardVMRunner {
                                        return;
                                }
 
+                               if (THROW) {
+                                       throw new BindException("test"); //$NON-NLS-1$
+                               }
                                connector.startListening(map);
 
                                p = exec(cmdLine, cmdDetails.getWorkingDir(), cmdDetails.getEnvp(), config.isMergeOutput());
```

Results in a pass and this in the log:

```
WARNING: Using incubator modules: jdk.incubator.vector
!SESSION 2026-04-15 20:47:38.806 -----------------------------------------------
eclipse.buildId=4.40.0.I20260414-1800
java.version=21.0.10
java.vendor=Arch Linux
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Framework arguments:  -version 3 -port 40873 -testLoaderClass org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader -loaderpluginname org.eclipse.jdt.junit4.runtime -test org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest:test547462_BooleanExpression -application org.eclipse.pde.junit.runtime.uitestapplication -product org.eclipse.sdk.ide -testpluginname org.eclipse.jdt.debug.tests
Command-line arguments:  -os linux -ws gtk -arch x86_64 -consoleLog -version 3 -port 40873 -testLoaderClass org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader -loaderpluginname org.eclipse.jdt.junit4.runtime -test org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest:test547462_BooleanExpression -application org.eclipse.pde.junit.runtime.uitestapplication -product org.eclipse.sdk.ide -data /data/workspaces/contributor_workspace/../junit-workspace -dev file:///data/workspaces/contributor_workspace/.metadata/.plugins/org.eclipse.pde.core/pde-junit/dev.properties -os linux -ws gtk -arch x86_64 -consoleLog -testpluginname org.eclipse.jdt.debug.tests

!ENTRY org.eclipse.equinox.app 0 0 2026-04-15 20:47:40.239
!MESSAGE Product org.eclipse.sdk.ide could not be found.
WARNING: Annotation classes from the 'javax.inject' or 'javax.annotation' package found.
It is recommended to migrate to the corresponding replacements in the jakarta namespace.
The Eclipse E4 Platform will remove support for those javax-annotations in a future release.
To suppress this warning, set the VM property: -Declipse.e4.inject.javax.warning=false
To disable processing of 'javax' annotations entirely, set the VM property: -Declipse.e4.inject.javax.disabled=true


!ENTRY org.eclipse.e4.ui.css.swt.theme 2 0 2026-04-15 20:47:41.009
!MESSAGE Theme 'org.eclipse.e4.ui.css.theme.e4_default' not found; falling back to default theme.
20:47:42.648 SETTING UP TESTS
20:47:42.649 SETUP ExpressionEvalTest.test547462_BooleanExpression

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.025
!MESSAGE VM added java-21-openjdk, location: /usr/lib/jvm/java-21-openjdk

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.161
!MESSAGE VM added java-26-openjdk, location: /usr/lib/jvm/java-26-openjdk

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.162
!MESSAGE VM added java-11-openjdk, location: /usr/lib/jvm/java-11-openjdk

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.162
!MESSAGE VM added java-25-openjdk, location: /usr/lib/jvm/java-25-openjdk

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.162
!MESSAGE VM added java-8-openjdk, location: /usr/lib/jvm/java-8-openjdk

!ENTRY org.eclipse.jdt.debug 1 0 2026-04-15 20:47:43.163
!MESSAGE VM added java-17-openjdk, location: /usr/lib/jvm/java-17-openjdk

!ENTRY org.eclipse.jdt.debug.tests 4 0 2026-04-15 20:47:44.993
!MESSAGE Could not bind socket for debugging, retry: 1
!STACK 1
org.eclipse.core.runtime.CoreException: Cannot connect to VM
	at org.eclipse.jdt.launching.AbstractVMRunner.abort(AbstractVMRunner.java:56)
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:462)
	at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1500)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1476)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1684)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1639)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1611)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1598)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3022)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:126)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:131)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1051)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:191)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:615)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:563)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1415)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1387)
Caused by: java.net.BindException: test
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:338)
	... 55 more
!SUBENTRY 1 org.eclipse.jdt.launching 4 120 2026-04-15 20:47:44.994
!MESSAGE Cannot connect to VM
!STACK 0
java.net.BindException: test
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:338)
	at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1500)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1476)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1684)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1639)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1611)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1598)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3022)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:126)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:131)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1051)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:191)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:615)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:563)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1415)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1387)

!ENTRY org.eclipse.jdt.debug.tests 4 0 2026-04-15 20:47:44.999
!MESSAGE Could not bind socket for debugging, retry: 2
!STACK 1
org.eclipse.core.runtime.CoreException: Cannot connect to VM
	at org.eclipse.jdt.launching.AbstractVMRunner.abort(AbstractVMRunner.java:56)
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:462)
	at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1500)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1476)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1684)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1639)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1611)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1598)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3022)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:126)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:131)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1051)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:191)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:615)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:563)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1415)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1387)
Caused by: java.net.BindException: test
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:338)
	... 55 more
!SUBENTRY 1 org.eclipse.jdt.launching 4 120 2026-04-15 20:47:44.999
!MESSAGE Cannot connect to VM
!STACK 0
java.net.BindException: test
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:338)
	at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1500)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1476)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1684)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1639)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1611)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1598)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3022)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:126)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:131)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1051)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:191)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:615)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:563)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1415)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1387)
20:47:46.329 TDOWN ExpressionEvalTest.test547462_BooleanExpression

!ENTRY org.eclipse.core.resources 2 10035 2026-04-15 20:47:47.083
!MESSAGE The workspace will exit with unsaved changes in this session.
```

Having the code in `StandardVMDebugger` always throw results in this failure:

```
org.eclipse.core.runtime.CoreException: Cannot connect to VM
	at org.eclipse.jdt.launching.AbstractVMRunner.abort(AbstractVMRunner.java:56)
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:462)
	at org.eclipse.jdt.launching.JavaLaunchDelegate.launch(JavaLaunchDelegate.java:176)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:778)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1500)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchAndWait(AbstractDebugTest.java:1476)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1684)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1639)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1611)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.launchToBreakpoint(AbstractDebugTest.java:1598)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.debugWithBreakpoint(ExpressionEvalTest.java:116)
	at org.eclipse.jdt.debug.tests.eval.ExpressionEvalTest.test547462_BooleanExpression(ExpressionEvalTest.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:3022)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:126)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:131)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5051)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1051)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:684)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:191)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:615)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:563)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1415)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1387)
Caused by: java.net.BindException: test
	at org.eclipse.jdt.internal.launching.StandardVMDebugger.run(StandardVMDebugger.java:338)
	... 55 more
```

--- a/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java
+++ b/org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/AbstractDebugTest.java
@@ -19,6 +19,7 @@ package org.eclipse.jdt.debug.tests;
 import java.io.File;
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.net.BindException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -177,6 +178,7 @@ import junit.framework.TestCase;
 @SuppressWarnings("deprecation")
 public abstract class AbstractDebugTest extends TestCase implements  IEvaluationListener {
 
+	private static final int SOCKET_BIND_ERROR_MAX_RETRIES = 3;
 	private static boolean setupFirstTest = false;
 
 	public static final String MULTI_OUTPUT_PROJECT_NAME = "MultiOutput";
@@ -1490,13 +1492,7 @@ public abstract class AbstractDebugTest extends TestCase implements  IEvaluation
 	 *             if the event is never received.
 	 */
 	protected Object launchAndWait(ILaunchConfiguration configuration, String mode, DebugEventWaiter waiter, boolean register) throws CoreException {
-		ILaunch launch;
-		try {
-			launch = configuration.launch(mode, new TimeoutMonitor(DEFAULT_TIMEOUT), false, register);
-		} catch (Throwable t) {
-			logProcessConsoleContents();
-			throw t;
-		}
+		ILaunch launch = launchConfig(configuration, mode, register);
 		Object suspendee= waiter.waitForEvent();
 		if (suspendee == null) {
 			StringBuilder buf = new StringBuilder();
@@ -1538,8 +1534,6 @@ public abstract class AbstractDebugTest extends TestCase implements  IEvaluation
 		}
 		return suspendee;
 	}
-
-
 
 	/**
 	 * Launches the type with the given name, and waits for a
@@ -3222,6 +3216,28 @@ public abstract class AbstractDebugTest extends TestCase implements  IEvaluation
 		String detailed = message + " " + vm.getName() + ", location: " + vm.getInstallLocation();
 		IStatus status = new Status(IStatus.INFO, JDIDebugPlugin.getUniqueIdentifier(), detailed, null);
 		JDIDebugPlugin.log(status);
+	}
+
+	private static ILaunch launchConfig(ILaunchConfiguration configuration, String mode, boolean register) throws CoreException {
+		ILaunch launch = null;
+		for (int retry = 1; retry <= SOCKET_BIND_ERROR_MAX_RETRIES; ++retry) {
+			try {
+				launch = configuration.launch(mode, new TimeoutMonitor(DEFAULT_TIMEOUT), false, register);
+				break;
+			} catch (CoreException e) {
+				Throwable cause = e.getStatus().getException();
+				if (retry < SOCKET_BIND_ERROR_MAX_RETRIES && cause instanceof BindException) {
+					// port might be in use between checking for a free port and launching, try a few times
+					DebugPlugin.log(Status.error("Could not bind socket for debugging, retry: " + retry, e));
+				} else {
+					throw e;
+				}
+			} catch (Throwable t) {
+				logProcessConsoleContents();
+				throw t;
+			}
+		}
+		return launch;
 	}
 
 	private static void logProcessConsoleContents() {

